[JSC] Cache FunctionExecutable from `new Function` constructor call with one-entry cache

Constellation opened the pull request:
#### 2581d153d8bee7078509ef4b5a27ad2a7090bb72
<pre>
[JSC] Cache FunctionExecutable from `new Function` constructor call with one-entry cache
<a href="https://bugs.webkit.org/show_bug.cgi?id=286233">https://bugs.webkit.org/show_bug.cgi?id=286233</a>
<a href="https://rdar.apple.com/143208554">rdar://143208554</a>

Reviewed by Yijia Huang.

This patch introduces a simple one-entry cache for `new Function`
constructor&apos;s FunctionExecutable. We also handle two-arguments `new
Function` calls efficiently.

* LayoutTests/inspector/debugger/sourceURL-repeated-identical-executions-expected.txt:
* LayoutTests/inspector/debugger/sourceURL-repeated-identical-executions.html:
* LayoutTests/js/script-tests/regress-150513.js:
(test):
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/parser/SourceCodeKey.h:
(JSC::SourceCodeKey::functionConstructorParametersEndPosition const):
* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::functionConstructorPrefix):
(JSC::stringifyFunction):
(JSC::constructFunction):
(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionConstructor.h:
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::fromGlobalCode):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryGetCachedFunctionExecutableForFunctionConstructor):
(JSC::JSGlobalObject::cachedFunctionExecutableForFunctionConstructor):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/SourceOrigin.h:

Canonical link: <a href="https://commits.webkit.org/289677@main">https://commits.webkit.org/289677@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b3673e6fe0fcbeafd1282499f1f1ae9eaf5337fc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/85857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/5489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/40232 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/90882 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/36786 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/5721 "Built successfully") | [❌ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/13491 "Hash b3673e6f for PR 39270 does not build (failure)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/66655 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/24449 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/88861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/4378 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/77882 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/46944 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/4237 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/32162 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/35863 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/78815 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/74880 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/33016 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/92601 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/84799 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/13123 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/13491 "Hash b3673e6f for PR 39270 does not build (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/75407 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/13333 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/73726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/74552 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/18764 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/17201 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/6016 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13666 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/13154 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/18510 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/107206 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/12935 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25824 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/16369 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/14722 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
```diff
@@ -66,23 +66,24 @@ void FunctionConstructor::finishCreation(VM& vm, FunctionPrototype* functionProt
-        break;
+        return "async function* "_s;
     }
+    return ASCIILiteral { };
```
nits: should add `ASSERT_NOT_REACHED();` ?

Constellation replied:
Changed.

justinmichaud commented on Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
```diff
@@ -129,15 +131,25 @@ DEFINE_VISIT_OUTPUT_CONSTRAINTS(FunctionExecutable);
-    JSObject*& exception, int overrideLineNumber, std::optional<int> functionConstructorParametersEndPosition)
+    JSObject*& exception, int overrideLineNumber, std::optional<int> functionConstructorParametersEndPosition, FunctionConstructionMode functionConstructionMode)
 {
+    if (overrideLineNumber == -1) {
```
Can we get a named constant for this?

hyjorc1 replied:
Yeah, maybe we can use this `constexpr int NO_OVERRIDE_LINE_NUMBER = -1;`?

hyjorc1 commented on Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
```diff
@@ -129,15 +131,25 @@ DEFINE_VISIT_OUTPUT_CONSTRAINTS(FunctionExecutable);
 
-    return unlinkedExecutable->link(globalObject->vm(), nullptr, source, overrideLineNumber);
+    auto* executable = unlinkedExecutable->link(globalObject->vm(), nullptr, source, overrideLineNumber);
+    if (overrideLineNumber == -1) {
```
Ditto

hyjorc1 reviewed (approved):
r=me

Constellation committed:
[JSC] Cache FunctionExecutable from `new Function` constructor call with one-entry cache
https://bugs.webkit.org/show_bug.cgi?id=286233
rdar://143208554

Reviewed by Yijia Huang.

This patch introduces a simple one-entry cache for `new Function`
constructor's FunctionExecutable. We also handle two-arguments `new
Function` calls efficiently.

* LayoutTests/inspector/debugger/sourceURL-repeated-identical-executions-expected.txt:
* LayoutTests/inspector/debugger/sourceURL-repeated-identical-executions.html:
* LayoutTests/js/script-tests/regress-150513.js:
(test):
* Source/JavaScriptCore/parser/ParserModes.h:
* Source/JavaScriptCore/parser/SourceCodeKey.h:
(JSC::SourceCodeKey::functionConstructorParametersEndPosition const):
* Source/JavaScriptCore/runtime/CodeCache.cpp:
(JSC::CodeCache::getUnlinkedGlobalFunctionExecutable):
* Source/JavaScriptCore/runtime/FunctionConstructor.cpp:
(JSC::functionConstructorPrefix):
(JSC::stringifyFunction):
(JSC::constructFunction):
(JSC::constructFunctionSkippingEvalEnabledCheck):
* Source/JavaScriptCore/runtime/FunctionConstructor.h:
* Source/JavaScriptCore/runtime/FunctionExecutable.cpp:
(JSC::FunctionExecutable::fromGlobalCode):
* Source/JavaScriptCore/runtime/FunctionExecutable.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::tryGetCachedFunctionExecutableForFunctionConstructor):
(JSC::JSGlobalObject::cachedFunctionExecutableForFunctionConstructor):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/SourceOrigin.h:

Canonical link: https://commits.webkit.org/289677@main
